### PR TITLE
Stateless Sequencer amnesia fix

### PR DESCRIPTION
**Pull request by zalokhan**
No longer persisting state in the sequencer.
Added randomized test for recovering sequencer state.

### CODE REVIEW

File: test/src/test/java/org/corfudb/integration/AbstractIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.concurrent.Executors;
+
+/**
+ * Created by zlokhandwala on 4/28/17.
+ */
+public class AbstractIntegrationTest extends AbstractCorfuTest {
+
+    private static final String TEST_HOST = "localhost";
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/integration/AbstractIntegrationTest.java]: What if you need multiple servers? having static ports doesnt seem like a good idea. We can address that later though.

File: test/src/test/java/org/corfudb/integration/ServerRestartTest.java [hunk 2]
```diff
@@ -1,160 +1,73 @@
...
 
     /**
      * Test recovery with a failed over server and a new client.
      *
      * @throws Exception
      */
     @Test
-    public void testSequencerRecoveryFailoverServerAndClient() throws Exception {
-        setUp();
+    public void testRecoveryFailoverServerAndClient() throws Exception {
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/integration/ServerRestartTest.java]: It seems like all these tests are just a subset of testRandomizedRecovery. Can you clobber them?

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 3]
```diff
@@ -509,41 +493,4 @@ private void handleAllocation(CorfuPayloadMsg<TokenRequest> msg,
     }
 
     private static final int globalTokenBatchSize = 100;
-
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java]: remove previous line as well

File: test/src/test/java/org/corfudb/integration/AbstractIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
+package org.corfudb.integration;
```
Comment (dahliamalkhi) [inline test/src/test/java/org/corfudb/integration/AbstractIntegrationTest.java]: cool, this is a userful utility class for integration tests.

**Comment by Maithem** [comment]
Please make sure that you address the comments in another patch.

File: test/src/test/java/org/corfudb/integration/ServerRestartTest.java [hunk 5]
```diff
@@ -163,87 +76,174 @@ public void testDoubleFailoverRecovery() throws Exception {
...
         map.put("a", expectedValue);
 
-        map = createMap(createRuntime());
+        shutdownCorfuServer(corfuServerProcess);
 
-        assertThat(map.get("a")).isEqualTo(expectedValue);
+        assert !corfuServerProcess.isAlive();
+
+        corfuServerProcess = runCorfuServer();
+
```
Comment (dahliamalkhi) [inline test/src/test/java/org/corfudb/integration/ServerRestartTest.java]: add comment here: do we expect some delay until reconfig or reconnect or both ?

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java (hunk 3 above)
Comment (zalokhan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java]: Ok Done.

**Comment by zalokhan** [comment]
@Maithem Yes, I will address the comments in the next patch

File: test/src/test/java/org/corfudb/integration/ServerRestartTest.java (hunk 5 above)
Comment (zalokhan) [inline test/src/test/java/org/corfudb/integration/ServerRestartTest.java]: Yes, we do expect slight delay. I'll add a comment.

File: test/src/test/java/org/corfudb/integration/ServerRestartTest.java [hunk 6]
```diff
@@ -1,160 +1,73 @@
...
         final int expectedValue = 5;
 
         Integer previous = map.get("a");
         assertThat(previous).isNull();
         map.put("a", expectedValue);
 
+        // Shut down and assert node is down.
         shutdownCorfuServer(corfuServerProcess);
-
         assert !corfuServerProcess.isAlive();
```
Comment (zalokhan) [inline test/src/test/java/org/corfudb/integration/ServerRestartTest.java]: I'll make the change.

File: test/src/test/java/org/corfudb/integration/ServerRestartTest.java (hunk 2 above)
Comment (zalokhan) [inline test/src/test/java/org/corfudb/integration/ServerRestartTest.java]: I'll reduce redundancy.

**Comment by medhavidhawan** [comment]
If you are going to do a second pass, please add facility to dump the sequence of actions when there is a failure. In other words please keep an array of actions (as defined by a test) as a trace of the test and dump it for analysis if there is a failure. You could also dump the seed of the Random number generator. I think we can build on this abstraction and can have a lot of smoke tests for end to end functionality. It can also be a good facilitator in fuzz testing. Please ping me for your second iteration, we can discuss some ideas.

**Pull request merged**